### DISCUSSION TITLE
Error launching App on MacOS with Shared Runtime

### DISCUSSION THREAD

**Discussion by KarlSchumann** [airsdk/Adobe-Runtime-Support#1165] [Q&A]
We have an application that we are hoping to distribute leveraging the shared runtime as that is currently how installation and updates are setup to work and have worked with previous versions of AIR.

On MacOS we are running into an issue where our app that was built using  33.1.1.575 and using the 33.1.1.533 version of the shared runtime wont open on Mac and we receive the following error: You do not have permission to open the application {app-name}"

![Error](https://user-images.githubusercontent.com/26549853/133663327-6af7003a-9a8b-4b51-8887-96a2b141c4e9.png)

Is this related to the known issue dealing with publisher verification? Is there anyway around this if we are using a code signing certificate from digicert? Is resolving the publisher verification issue on the roadmap to be resolved in a future release or is that something that will always exist with the shared runtime?

The same build is working correctly on Windows and our previous version of our app that was built with AIR 32 opens correctly when using the HARMAN version of the runtime.

Any help with this would be appreciated.

**Comment by itlancer** [comment]
I have the same problem: https://github.com/airsdk/Adobe-Runtime-Support/issues/990#issuecomment-883413067
Now I switched to use AIR application as bundle (instead of using shared runtime) and it works without such issues: https://github.com/airsdk/Adobe-Runtime-Support/issues/990#issuecomment-883456148

**Reply by KarlSchumann** [reply]
Thank you for the response, the bundle option might be the path we need to head down. Will have to try and get an environment setup on a Mac as currently have been doing most of our development and building on Windows.

**Comment by ajwfrost** [comment]
@KarlSchumann It's not about the publisher verification, I think, as we get a different error/display when that causes a problem. I was going to say that it's related to the security/signing on new macOS versions, but the fact you're saying that it will install okay if you're using an old version of your .air file is somewhat curious.

Are you able to send us the application - as a .air file - both as the AIR32 version, and the AIR33.1 version? We can run this through with some logs and also check the outcome to see what's different.
https://transfer.harman.com/requests/y1KCdE6Gt8mR2pKp6d6Lan

thanks

**Reply by KarlSchumann** [reply]
Thanks Andrew that would be great if you could take a look. I just submitted both, hopefully they went through okay. Let me know if there is anything else I can provide to help.

**Reply by ajwfrost** [reply]
@KarlSchumann well for me on Big Sur, I got that same error on both of the .air files that you provided.. and I could only get rid of it after I code-signed the installed applications. The 2021 .air file worked fine when I installed it on a Mojave machine.

On my Big Sur machine here, I only have the ability to run apps that are signed with either a Developer ID Application certificate, or that have been installed from the App Store. So when the AIR framework 'installs' the .air file, it's basically copying files into the right place to make it look like an application, but of course we have no way at that point of code-signing it....

The only solutions to this that I can think of will all involve using a Mac. The simplest option is to use the 'bundle' type so that you create an application folder and can sign this, and then distribute it already signed as a 'normal' application. Keeping it as a .air file though, it may be possible for us to come up with a solution that embeds a signed binary into the .air package in a way that's ignored on Windows but is used on a Mac to restore the signature.. but this is going to be a bit of work.

So I'm afraid for the short term, you're best off using a bundle, and signing it using a Developer ID Application certificate..

thanks

**Reply by KarlSchumann** [reply]
@ajwfrost thank you for taking a look at this, I appreciate you taking the time to look at our specific situation and the feedback.

I now realize I had installed the 2019 version when I was still using Adobe's v32 of the runtime. I downgraded to retest and when I install the provided 2019 version with the 32 version of the AIR Runtime on Big Sur I do not receive the "You do not have permission" issue. Of course for that to work the quarantine bit on the AIR Installer needs to be removed which is an unfortunate step to have customers go through. 

Is there a reason why the code signing seemingly works correctly with the 32 version of the shared runtime and doesn't require extra steps  vs the 33 version of the shared runtime?

Our app took advantage of the built in application updates AIR provided which worked pretty well for cross platform and allowed us to not have to build separate installers for Windows and Mac, but it seems like if we are going to continue to use AIR that might be the path we have to head down.

**Reply by KarlSchumann** [reply]
If I then upgrade to HARMAN's release of the shared Runtime I will see the splash screen since that app wasn't built with the license, but it will still open correctly so it seems like there is a difference between the way the v32 and v33 are installing the apps and dealing with the code signing.

**Reply by KarlSchumann** [reply]
@ajwfrost Do happen to have any more information to share on why the new shared runtime isn't handling the code signing properly with the same build that the older shared runtime was? Is this something on the roadmap to fix or is there something I might be doing incorrectly to build with the newer HARMAN sdk? Or is the only path forward to use the captive builds?
